include/LLVMSPIRVOpts.h

```diff
@@ -99,12 +102,63 @@ class TranslatorOpts {
 
   void enableGenArgNameMD() { GenKernelArgNameMD = true; }
 
+  // Stores values for specialization constants. Potentially, can be used for
+  // implementation of clSetProgramSpecializationConstant API.
```

Comment:
clSetProgramSpecializationConstant API is seperate interface from clBuildProgram. 
Lets imagine potential controlflow and decide whether such function is fully reasonable in this case.
User calls clSetProgramSpecializationConstant from host code, then Runtime needs to verify whether specified specialization constant is valid with specialization constant within the kernel (lets remember that clBuildProgram does not have to be called yet, so we have no kernel compiled). At this moment Runtime/Compiler needs to parse and only parse the kernel(it does not have to translate it) to validate whether specialization constant (passed as an arguments to API call) is correct. 
In my opinion, we should implement the function parsing SPIRV module and returning specialization constants within it. It would be more helpful for OpenCL Runtimes as they could just call this function and use the result to validate user-specified specialization constant.

Now, lets imagine that OpenCL Runtime calls clBuildProgram. In my opinion your function (addSpecConst) is reasonable in this case only, as Runtime/Compiler needs to update previously gathered (during clSetProgramSpecializationConstant) specialization constants.

Please correct me if my way of thinking is unreasonable. Thanks.

Reply:
I think your reasoning makes sense and it is related to my TODO a couple lines below.
It seems we do need an API to check spec constants passed to clSetProgramSpecializationConstant against the SPIR-V module. I'll update the patch.